README.md
# bpcalc
Whirlwind breakpoint calculator for Diablo 3

## Overview
bpcalc is a command-line utility for calculating Whirlwind and Dust Devil breakpoints for Barbarians in Diablo 3. The program will also show the required `increased attack speed` to reach the previous or next breakpoint.

Read: it shows wasted IAS as well as how much you need to reach the next breakpoint.

`Note:` binary packages were built for Windows 10 x64 only. You will need to download and install the Visual C++ 2015 runtime located here in order to run bpcalc: [http://www.microsoft.com/en-us/download/details.aspx?id=48145](http://www.microsoft.com/en-us/download/details.aspx?id=48145)

## Usage
Using `bpcalc` is really easy, simply type `bpcalc [options]` in a command prompt where options correspond to the following options:
```
    Option       Argument     Description
    ------       --------     -----------
    -bk                       Bul-Kathos set's 30% IAS multiplier
    -frenzied                 Frenzied Shrine's 25% IAS bonus
    -gogok                    Gogok of Swiftness's 15% IAS bonus
    -ias         <number>     Increased attack speed (shown on character sheet)
    -pe          <enemies>    Pain Enhancer's 3% IAS bonus per bleeding enemy
    -wep         <number>     Weapon speed
    -wotb                     Wrath of the Berserker's 25% IAS bonus
```
An example might look something like this, assuming WoTB is up and you have BK weapons with 7% native IAS on them and 30% IAS from gear: `bpcalc -bk -ias 30 -wep 1.39` which would output the following:
```
Weapon speed            IAS                   APS
------------            ---                   ---
1.39                    30.0%                 2.35

Ticks per second        IAS to previous BP    IAS to next BP
----------------        ------------------    --------------
7.50                    22.9% (-7.1%)         38.4% (+8.4%)

Tornadoes per second    IAS to previous BP    IAS to next BP
--------------------    ------------------    --------------
5.00                    27.6% (-2.4%)         38.4% (+8.4%)
```

## Building from source
To build from the source, run `git clone https://github.com/snailsoft/bpcalc.git` and then open the sln file.

`Note:` Visual Studio 2015 is required to build bpcalc from the source files.


changelog.txt
bpcalc - Whirlwind Breakpoint Calculator
Copyright(C) 2015 Snailsoft <https://github.com/snailsoft/bpcalc/>

------------------------------------------------------------------------------

Version 1.0.0
- Initial release


src/bpcalc.hpp
/*
 * Copyright (C) 2015 Snailsoft <https://github.com/snailsoft/bpcalc/>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

#include <iostream>
#include <whirlwind.hpp>

bool get_argument(int &argc, char **&argv, int &argument, float &value);


src/version.hpp
/*
 * Copyright (C) 2015 Snailsoft <https://github.com/snailsoft/bpcalc/>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

#define BPCALC_AUTHOR			"Snailsoft"
#define BPCALC_DESCRIPTION		"Whirlwind Breakpoint Calculator"
#define BPCALC_NAME				"bpcalc"
#define BPCALC_VERSION			"1.0.0"
#define BPCALC_VERSION_NUMBER	1, 0, 0


src/whirlwind.cpp
/*
 * Copyright (C) 2015 Snailsoft <https://github.com/snailsoft/bpcalc/>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

#include <whirlwind.hpp>

float Whirlwind::get_attacks_per_second()
{
	return (this->weapon_speed * (1.0f + this->increased_attack_speed)) * (this->bul_kathos ? 1.3f : 1.0f);
}

float Whirlwind::get_breakpoint(float divider, bool next)
{
	auto breakpoint = 0.0f;
	auto ias = this->increased_attack_speed;
	auto tps = this->get_ticks_per_second(divider);

	while(tps == this->get_ticks_per_second(divider))
		this->increased_attack_speed += (next ? 0.0001f : -0.0001f);

	breakpoint = this->increased_attack_speed - (next ? 0.0f : 0.001f);
	this->increased_attack_speed = ias;

	return breakpoint;
}

float &Whirlwind::get_increased_attack_speed()
{
	return this->increased_attack_speed;
}

float Whirlwind::get_ticks_per_second(float divider)
{
	return 60.0f / floor(divider / this->get_attacks_per_second());
}

float &Whirlwind::get_weapon_speed()
{
	return this->weapon_speed;
}

Whirlwind::Whirlwind()
{
	this->bul_kathos = false;
	this->increased_attack_speed = 0.0f;
	this->weapon_speed = 0.0f;
}

Whirlwind::Whirlwind(float &weapon_speed, float &increased_attack_speed, bool &bul_kathos)
{
	this->bul_kathos = bul_kathos;
	this->increased_attack_speed = increased_attack_speed;
	this->weapon_speed = weapon_speed;
}


src/whirlwind.hpp
/*
 * Copyright (C) 2015 Snailsoft <https://github.com/snailsoft/bpcalc/>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

#include <cmath>

class Whirlwind
{
	private:
		bool bul_kathos;

		float increased_attack_speed;
		float weapon_speed;

	public:
		float get_attacks_per_second();
		float get_breakpoint(float divider, bool next = false);
		float &get_increased_attack_speed();
		float get_ticks_per_second(float divider);
		float &get_weapon_speed();

		Whirlwind();
		Whirlwind(float &weapon_speed, float &increased_attack_speed, bool &bul_kathos);
};


src/bpcalc.cpp
/*
* Copyright (C) 2015 Snailsoft <https://github.com/snailsoft/bpcalc/>
*
* This program is free software; you can redistribute it and/or modify it
* under the terms of the GNU General Public License as published by the
* Free Software Foundation; either version 2 of the License, or (at your
* option) any later version.
*
* This program is distributed in the hope that it will be useful, but WITHOUT
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
* FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
* more details.
*
* You should have received a copy of the GNU General Public License along
* with this program. If not, see <http://www.gnu.org/licenses/>.
*/

#include <bpcalc.hpp>
#include <version.hpp>

bool get_argument(int &argc, char **&argv, int &argument, float &value)
{
	if((argument + 1) <= (argc - 1))
	{
		value = static_cast<float>(atof(argv[++argument]));

		return true;
	}

	return false;
}

int main(int argc, char *argv[])
{
	std::cout << BPCALC_NAME << " - " << BPCALC_DESCRIPTION << std::endl;
	std::cout << "Version " << BPCALC_VERSION << std::endl;
	std::cout << std::endl;
	std::cout << "Copyright(C) 2015 " << BPCALC_AUTHOR << " <https://github.com/snailsoft/bpcalc/>" << std::endl;
	std::cout << std::endl;

	if(argc > 1)
	{
		auto bul_kathos = false;
		auto increased_attack_speed = 0.0f;
		auto weapon_speed = 0.0f;

		for(int i = 1; i < argc; ++i)
		{
			if(!strcmp(argv[i], "-bk"))
				bul_kathos = true;

			else if(!strcmp(argv[i], "-frenzied") || !strcmp(argv[i], "-wotb"))
				increased_attack_speed += 0.25f;

			else if(!strcmp(argv[i], "-gogok"))
				increased_attack_speed += 0.15f;

			else if(!strcmp(argv[i], "-ias"))
			{
				float ias = 0.0f;

				if(!get_argument(argc, argv, i, ias))
				{
					std::cout << "Error: Incorrect usage of -ias <number>" << std::endl;

					return 0;
				}

				increased_attack_speed += ias * 0.01f;
			}
			else if(!strcmp(argv[i], "-pe"))
			{
				float enemies = 0.0f;

				if(!get_argument(argc, argv, i, enemies))
				{
					std::cout << "Error: Incorrect usage of -pe <enemies>" << std::endl;

					return 0;
				}

				increased_attack_speed += enemies * 0.03f;
			}
			else if(!strcmp(argv[i], "-wep"))
			{
				if(!get_argument(argc, argv, i, weapon_speed))
				{
					std::cout << "Error: Incorrect usage of -wep <number>" << std::endl;

					return 0;
				}
			}
			else
				std::cout << "Warning: Invalid option: " << argv[i] << std::endl;
		}

		Whirlwind ww(weapon_speed, increased_attack_speed, bul_kathos);

		std::cout << "Weapon speed            IAS                   APS" << std::endl;
		std::cout << "------------            ---                   ---" << std::endl;

		printf("%.02f\t                %.01f%%\t              %.02f\n\n", ww.get_weapon_speed(), ww.get_increased_attack_speed() * 100.0f, ww.get_attacks_per_second());
		
		std::cout << "Ticks per second        IAS to previous BP    IAS to next BP" << std::endl;
		std::cout << "----------------        ------------------    --------------" << std::endl;
		
		printf("%.02f\t                %.01f%% (-%.01f%%)\t      %.01f%% (+%.01f%%)\n\n", ww.get_ticks_per_second(20.0f), ww.get_breakpoint(20.0f) * 100.0f, (ww.get_increased_attack_speed() - ww.get_breakpoint(20.0f)) * 100.0f, ww.get_breakpoint(20.0f, true) * 100.0f, (ww.get_breakpoint(20.0f, true) - ww.get_increased_attack_speed()) * 100.0f);

		std::cout << "Tornadoes per second    IAS to previous BP    IAS to next BP" << std::endl;
		std::cout << "--------------------    ------------------    --------------" << std::endl;

		printf("%.02f\t                %.01f%% (-%.01f%%)\t      %.01f%% (+%.01f%%)\n", ww.get_ticks_per_second(30.0f), ww.get_breakpoint(30.0f) * 100.0f, (ww.get_increased_attack_speed() - ww.get_breakpoint(30.0f)) * 100.0f, ww.get_breakpoint(30.0f, true) * 100.0f, (ww.get_breakpoint(30.0f, true) - ww.get_increased_attack_speed()) * 100.0f);
	}
	else
	{
		std::cout << "Usage:" << std::endl;
		std::cout << "    " << argv[0] << " [options]" << std::endl;
		std::cout << std::endl;
		std::cout << "Options:" << std::endl;
		std::cout << "    Option       Argument     Description" << std::endl;
		std::cout << "    ------       --------     -----------" << std::endl;
		std::cout << "    -bk                       Bul-Kathos set's 30% IAS multiplier" << std::endl;
		std::cout << "    -frenzied                 Frenzied Shrine's 25% IAS bonus" << std::endl;
		std::cout << "    -gogok                    Gogok of Swiftness's 15% IAS bonus" << std::endl;
		std::cout << "    -ias         <number>     Increased attack speed (shown on character sheet)" << std::endl;
		std::cout << "    -pe          <enemies>    Pain Enhancer's 3% IAS bonus per bleeding enemy" << std::endl;
		std::cout << "    -wep         <number>     Weapon speed" << std::endl;
		std::cout << "    -wotb                     Wrath of the Berserker's 25% IAS bonus" << std::endl;
	}

	return 0;
}
